### PR TITLE
docs: Fir 12403 add linter rule requiring docs

### PR DESCRIPTION
**Pull request by stepansergeevitch**
Added new flake8 checkers for docstrings:
- flake8-docstrings - validate that every required object has a docsting.
- flake8-spellcheck - spelling checker for docs
- darglint - validated that docstings have all required items (args, return value, raised exceptions)

Added docstrings for client and common modules. Other modules will be covered in next PRs

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -73,6 +73,7 @@ warn_untyped_fields = True
 [flake8]
 max-line-length = 88
 per-file-ignores = __init__.py:F401
-ignore = E203, W503
+ignore = E203, W503, D100, D104, D105, D107, SC200
```
Comment (ptiurin) [inline setup.cfg:77]: I can't seem to find those error codes. Where are they listed?

File: src/firebolt/client/auth.py [hunk 2]
```diff
@@ -12,16 +12,26 @@
...
+    Attributes:
+        username (str): Username
+        password (str): Password
     """
 
     __slots__ = (
         "username",
         "password",
-        "api_url",
+        "_api_endpoint",
```
Comment (ptiurin) [inline src/firebolt/client/auth.py:34]: Love that we spotted this! However, I thought it would fail on runtime if an instance variable was not declared in __slots__ ? How was this working before?

File: src/firebolt/client/auth.py [hunk 3]
```diff
@@ -50,18 +68,40 @@ def __init__(
...
+            Optional[str]: Acquired token
+        """
         return self._token
 
     @property
     def expired(self) -> Optional[int]:
+        """Check if current token is expired.
+
+        Returns:
+            bool: True if expired, False otherwise
```
Comment (ptiurin) [inline src/firebolt/client/auth.py:92]: Looks like it does return boolean, but our type hint is wrong.

File: src/firebolt/client/client.py [hunk 4]
```diff
@@ -30,34 +32,51 @@ def __init__(
...
-    httpx.Client:
-
-    + (HttpxClient.__doc__ or "")
     """
 
     @cached_property
     def account_id(self) -> str:
+        """User account id.
+
+        If account_name is provided, returns it's id.
```
Comment (ptiurin) [inline src/firebolt/client/client.py:71]: I'd rephrase it to something like "If account_name was provided during the instantiation of this class...".
This way we avoid confusion that the docstring talks about a parameter that was removed from `account_id()` and the comment was not updated.

File: src/firebolt/client/resource_manager_hooks.py [hunk 5]
```diff
@@ -7,32 +7,53 @@
...
+    """Log HTTP requests.
+
+    Hook for an HTTP client
+
+    Args:
+        request (Request): Request to log
+    """
     logger.debug(
-        f"Request event hook: {request.method} {request.url} - Waiting for response"
+        "Request event hook: %s %s - Waiting for response", request.method, request.url
```
Comment (ptiurin) [inline src/firebolt/client/resource_manager_hooks.py:18]: Is this the best practice, as opposed to f-strings, according to the checker?

File: src/firebolt/common/token_storage.py [hunk 6]
```diff
@@ -126,9 +154,27 @@ def __init__(self, salt: str, username: str, password: str):
...
         return self.fernet.encrypt(bytes(data, encoding="utf-8")).decode("utf-8")
 
     def decrypt(self, data: str) -> Optional[str]:
+        """Decrypt encrypted data.
+
+        Args:
+            data (str): Encrypted data
+
+        Returns:
+            str: Decrypted data
```
Comment (ptiurin) [inline src/firebolt/common/token_storage.py:175]: probably need `optional` here as well, for consistency

File: setup.cfg (hunk 1 above)
Comment (stepansergeevitch) [reply setup.cfg:77]: Ds are related to darglint, SCs are related to spellcheck

File: src/firebolt/client/auth.py (hunk 2 above)
Comment (stepansergeevitch) [reply src/firebolt/client/auth.py:34]: This is actually a tricky one.
The reason seems to be that base Auth class doesn't have `__slots__` defined, and therefore it creates a `__dict__` entry, which makes `__slots__` redundant. I think in this case we can leave it just for documentation purposes

File: src/firebolt/client/auth.py (hunk 3 above)
Comment (stepansergeevitch) [reply src/firebolt/client/auth.py:92]: Hm, wondering how mypy didn't catch this

File: src/firebolt/client/client.py (hunk 4 above)
Comment (stepansergeevitch) [reply src/firebolt/client/client.py:71]: Makes sense!

File: src/firebolt/client/resource_manager_hooks.py (hunk 5 above)
Comment (stepansergeevitch) [reply src/firebolt/client/resource_manager_hooks.py:18]: Yes, this way it performs a lazy evaluation

**Review by ptiurin — APPROVED** [review]
LGTM, thanks for adding descriptions
